internal/clientutils/output.go

```diff
@@ -0,0 +1,113 @@
+package clientutils
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/glasskube/glasskube/api/v1alpha1"
+	"github.com/glasskube/glasskube/internal/controller/ctrlpkg"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/yaml"
+)
+
+type OutputFormat string
+
+const (
+	OutputFormatJSON OutputFormat = "json"
+	OutputFormatYAML OutputFormat = "yaml"
+)
+
+func Format(outputFormat OutputFormat, showAll bool, pkgs ...ctrlpkg.Package) (string, error) {
+	if !showAll {
+		for _, pkg := range pkgs {
+			switch p := pkg.(type) {
+			case *v1alpha1.ClusterPackage:
+				p.ObjectMeta = pruneExtraFields(p.ObjectMeta)
+			case *v1alpha1.Package:
+				p.ObjectMeta = pruneExtraFields(p.ObjectMeta)
+			default:
+				panic("unexpected package type")
+			}
+		}
+	}
+
+	var outputData []byte
+	for i := range pkgs {
+		if gvks, _, err := scheme.Scheme.ObjectKinds(pkgs[i]); err == nil && len(gvks) == 1 {
+			pkgs[i].SetGroupVersionKind(gvks[0])
+		} else {
+			return "", fmt.Errorf("failed to set GVK for package: %w\n", err)
+		}
+	}
+	switch outputFormat {
+	case OutputFormatJSON:
+		var err error
+		res := make([]map[string]any, 0)
+		for _, pkg := range pkgs {
+			if pkgAsMap, pruneErr := packageAsMap(pkg, showAll); pruneErr != nil {
+				return "", fmt.Errorf("failed to prune status during marshalling: %w", pruneErr)
+			} else {
+				res = append(res, pkgAsMap)
+			}
+		}
+		if outputData, err = json.MarshalIndent(res, "", "  "); err != nil {
+			return "", fmt.Errorf("failed to marshal output: %w\n", err)
+		}
+	case OutputFormatYAML:
+		var buffer bytes.Buffer
+		l := len(pkgs)
+		for _, pkg := range pkgs {
+			if p, pruneErr := packageAsMap(pkg, showAll); pruneErr != nil {
+				return "", fmt.Errorf("failed to prune status during marshalling: %w", pruneErr)
+			} else if data, err := yaml.Marshal(p); err != nil {
+				return "", fmt.Errorf("failed to marshal output: %w\n", err)
+			} else {
+				if l > 1 {
+					buffer.WriteString("---\n")
+				}
+				buffer.Write(data)
+			}
+		}
+		outputData = buffer.Bytes()
+	default:
+		return "", fmt.Errorf("unsupported output format: %v", outputFormat)
+	}
+
+	return string(outputData), nil
+}
+
+func pruneExtraFields(original v1.ObjectMeta) v1.ObjectMeta {
+	return v1.ObjectMeta{
+		Name:        original.Name,
+		Namespace:   original.Namespace,
+		Labels:      pruneNonGlasskubeKeys(original.Labels),
+		Annotations: pruneNonGlasskubeKeys(original.Annotations),
+	}
+}
+
+func pruneNonGlasskubeKeys(original map[string]string) map[string]string {
+	res := make(map[string]string)
+	for k, v := range original {
+		if strings.Contains(k, "packages.glasskube.dev") {
+			res[k] = v
+		}
+	}
+	return res
+}
```

Comment:
To be honest I still have mixed feelings about not showing the entire object by default, but even removing labels unless explicitly requested seems kind of excessive IMO :confused:

Reply:
I agree, but coming from the point of view of a "glasskube argocd starter", isn't it quite a stretch to assume the user needs a super simple do-it-all automation on the one hand, but on the other hand expect them to know which labels and which fields to put into the git repo, and which not? That goes from super easy to reasonably difficult very quick. But maybe that's how it is – I don't know how this could be resolved to be honest. What I would not do is to separate the behaviour between CLI and UI, because they should always behave the same way and offer the same experience.

Reply:
I understand that, but at the same time it makes any automation that relies on the output less obvious, e.g. `glasskube install foo -o json | jq "..."`.

I guess, all I'm saying is that this change is based on some big assumptions about how the software is used in practice and I kind of challenge the notion that the "gitops" use-case is the most common one. In the end, I think we can go forward with this change but we should keep our ears open for feedback and be prepared to change it again down the line. :shrug:

Reply:
Let's focus on the GitOps case for now